src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/InlineCompletion/InlineCompletionEndPoint.cs

```diff
@@ -0,0 +1,240 @@
+
+            return documentSnapshot;
+        }, cancellationToken).ConfigureAwait(false);
+
+        if (document is null)
+        {
+            return null;
+        }
+
+        var codeDocument = await document.GetGeneratedOutputAsync();
+        if (codeDocument.IsUnsupported())
+        {
+            return null;
+        }
+
+        var sourceText = await document.GetTextAsync();
+        var linePosition = new LinePosition(request.Position.Line, request.Position.Character);
+        var hostDocumentIndex = sourceText.Lines.GetPosition(linePosition);
+
+        var languageKind = _documentMappingService.GetLanguageKind(codeDocument, hostDocumentIndex);
+
+        // Map to the location in the C# document.
+        if (languageKind != RazorLanguageKind.CSharp ||
+            !_documentMappingService.TryMapToProjectedDocumentPosition(codeDocument, hostDocumentIndex, out var projectedPosition, out _))
+        {
+            _logger.LogInformation($"Unsupported location for {request.TextDocument.Uri}.");
+            return null;
+        }
+
+        request.Position = projectedPosition;
+        var response = await _languageServer.SendRequestAsync(LanguageServerConstants.RazorInlineCompletionEndpoint, request).ConfigureAwait(false);
+        var list = await response.Returning<InlineCompletionList>(cancellationToken);
+        if (list == null)
+        {
+            _logger.LogInformation($"Did not get any inline completions from delegation.");
+            return null;
+        }
+
+        var items = new List<InlineCompletionItem>();
+        var csharpDocOptions = codeDocument.GetCSharpDocument();
+        foreach (var item in list.Items)
+        {
+            var containsSnippet = item.TextFormat == InsertTextFormat.Snippet;
+            var range = item.Range ?? new Range { Start = projectedPosition, End = projectedPosition };
+
+            if (!_documentMappingService.TryMapFromProjectedDocumentRange(codeDocument, range, out var rangeInRazorDoc))
+            {
+                _logger.LogWarning($"Could not remap projected range {range} to razor document");
+                continue;
+            }
+
+            using var formattingContext = FormattingContext.Create(request.TextDocument.Uri, document, codeDocument, s_defaultFormattingOptions, _adhocWorkspaceFactory, isFormatOnType: true, automaticallyAddUsings: false);
+            if (!TryGetSnippetWithAdjustedIndentation(formattingContext, item.Text, hostDocumentIndex, out var newSnippetText))
+            {
+                continue;
+            }
+
+            var remappedItem = new InlineCompletionItem
+            {
+                Command = item.Command,
+                Range = rangeInRazorDoc,
+                Text = newSnippetText.ToString(),
+                TextFormat = item.TextFormat,
+            };
+            items.Add(remappedItem);
+        }
+
+        _logger.LogInformation($"Returning items.");
```

Comment:
If items.Count == 0 could we return `null` and log that we couldn't format / map things correctly?